Ensure `S3Service` is `STARTED` when creating client

DaveCTurner opened the pull request:
It's possible for another component to request a S3 client after the
node has started to shut down, and today the `S3Service` will dutifully
attempt to create a fresh client instance even if it is closed. Such
clients will then leak, resulting in test failures.

With this commit we refuse to create new S3 clients once the service has
started to shut down.

DaveCTurner committed:
Ensure `S3Service` is `STARTED` when creating client

It's possible for another component to request a S3 client after the
node has started to shut down, and today the `S3Service` will dutifully
attempt to create a fresh client instance even if it is closed. Such
clients will then leak, resulting in test failures.

With this commit we refuse to create new S3 clients once the service has
started to shut down.

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

arteam reviewed (approved):
LGTM!

Makes sense to avoid create new `SdkHttpClient` if `S3Service` is being shutdown

ywangd commented:
For my knowledge: Does this issue only matter for the v2 client where we explicitly build a `SdkHttpClient`? Or is it applicable to v1 client as well and we were just luck enough to not hit it in tests?

DaveCTurner commented:
In https://github.com/elastic/elasticsearch/pull/126843 we dropped the call to `IdleConnectionReaper.shutdown();` - the v2 SDK shuts down the connection reaper automatically when all the clients are closed and does not allow it to be shut down otherwise. Also we changed the lifecycle so that the `S3Service` is a proper `LifecycleComponent` in its own right, rather than something which closes when the `S3RepositoryPlugin` closes, which means it may now close after other services that are still trying to use it, particularly the `ObjectStoreService`.
